sphinx/ext/napoleon/docstring.py

```diff
@@ -872,6 +884,83 @@ def __init__(self, docstring: Union[str, List[str]], config: SphinxConfig = None
         self._directive_sections = ['.. index::']
         super().__init__(docstring, config, app, what, name, obj, options)
 
+    def _convert_type_spec(self, _type):
+        def recombine_set(tokens):
+            def combine_set(tokens):
+                in_set = False
+                set_items = []
+
+                for token in tokens:
+                    if token.startswith("{"):
```

Comment:
Does value_set support a literal of set or dict (ex. `{1, "2", {3}, {"key": 4}}`)?

Reply:
not sure, I didn't think of this so it probably won't. Is that valid `numpydoc`?

I suspect that if we want to support it, we somehow need to count opening and closing braces (or write/use a parser (json?))

Reply:
looking at this a bit more, the style reference is not really clear about this, it only says that this:
```
{'C', 'F', 'A'}
```
is valid. The `numpydoc` extension (which I'd assume is the reference implementation) actually doesn't enforce anything, it simply uses whatever is after the first colon as a type.

So since by parsing the types we are already enforcing a certain format, I guess we can decide if we want to enforce flattened sets

Reply:
Sorry for the lack of my words. I don't know what is the correct format of numpydoc. But users usually get mistakes. So we have to handle and let users know them.